pkg/nodeconfig/nodeconfig.go

```diff
@@ -288,6 +303,110 @@ func (nc *nodeConfig) Configure() error {
+		return err
+	}
+	return nil
+}
+
+// createFilesFromIgnition creates any files it can from ignition: kubelet CA cert, cloud-config file
+func (nc *nodeConfig) createFilesFromIgnition() error {
+	filesToTransfer := map[string]struct{}{
+		ignition.KubeletCACertPath: {},
+	}
+
+	ign, err := ignition.New(nc.client)
+	if err != nil {
+		return err
+	}
+	kubeletArgs, err := ign.GetKubeletArgs()
+	if err != nil {
+		return err
+	}
+	if _, ok := kubeletArgs[ignition.CloudConfigOption]; ok {
+		filesToTransfer[ignition.CloudConfigPath] = struct{}{}
+	}
+	// For each new file in the ignition file check if is a file we are interested in and, if so, decode its contents
+	for _, ignFile := range ign.GetFiles() {
+		if _, ok := filesToTransfer[ignFile.Node.Path]; ok {
+			if ignFile.Contents.Source == nil {
+				return errors.Errorf("could not process %s: File is empty", ignFile.Node.Path)
+			}
+			contents, err := url.QueryUnescape(*ignFile.Contents.Source)
+			if err != nil {
+				return errors.Wrapf(err, "could not decode %s", ignFile.Node.Path)
+			}
+			// Write to file on the Windows instance
+			fileName := filepath.Base(ignFile.Node.Path)
+			if err := nc.Windows.EnsureFileContent([]byte(contents), fileName, windows.K8sDir); err != nil {
+				return err
+			}
+			nc.log.V(1).Info("created", "file", windows.K8sDir+fileName)
+		}
+	}
+	return nil
+}
+
+// createBootstrapKubeconfig creates a kubeconfig for kubelet to initially communicate with the API server
+func (nc *nodeConfig) createBootstrapKubeconfig() error {
+	nodeBootstrapToken, err := nc.k8sclientset.CoreV1().Secrets(mcoNamespace).Get(context.TODO(), mcoBootstrapSecret,
+		meta.GetOptions{})
+	if err != nil {
+		return err
+	}
+	kubeconfig, err := generateKubeconfig(nodeBootstrapToken, nodeConfigCache.apiServerEndpoint)
+	if err != nil {
+		return err
+	}
+	kubeconfigData, err := json.Marshal(kubeconfig)
+	if err != nil {
+		return err
+	}
+
+	dir, fileName := splitWindowsPath(windows.BootstrapKubeconfig)
+	if err := nc.Windows.EnsureFileContent(kubeconfigData, fileName, dir); err != nil {
+		return err
+	}
+	nc.log.V(1).Info("created", "file", windows.BootstrapKubeconfig)
+	return nil
+}
+
+// createKubeletConf creates config file for kubelet, with Windows specific configuration.
+func (nc *nodeConfig) createKubeletConf(clusterDNS string) error {
+	kubeletConfig := generateKubeletConfiguration(clusterDNS)
+	kubeletConfigData, err := json.Marshal(kubeletConfig)
+	if err != nil {
+		return err
+	}
+	// Appending this option is needed here instead of in the kubelet configuration object. Otherwise, when marshalling,
+	// the empty value will be omitted, so it would end up being incorrectly populated at service start time.
+	enforceNodeAllocatable := "\"enforceNodeAllocatable\":[]}"
+	// Replace last character ('}') with comma
+	kubeletConfigData[len(kubeletConfigData)-1] = ','
+	kubeletConfigData = []byte(string(kubeletConfigData) + enforceNodeAllocatable)
```

Comment:
Consider using "..." as a suffix to append directly and avoid the string conversion.

```go
	kubeconfigData[len(kubeconfigData)-1] = ','
        kubeconfigData = append(kubeconfigData, enforceNodeAllocatable...)

```